### PR TITLE
LCOW: Implement `docker save`

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

This PR contains three small fixes around `docker save` and a DiffGetter implementation for the LCOW graphdriver. The smaller fixes are

- A check to ensure we don't nil de-reference saving an LCOW image in `image\tarexport\save.go`
- Making sure the operating system is passed through to `saveLayer`
- Making sure manifest.json file is written in Unix paths, otherwise the saved .tar from LCOW can't be loaded by a Linux daemon.

```
PS C:\> docker pull --platform=linux busybox
Using default tag: latest
latest: Pulling from library/busybox
d070b8ef96fc: Pull complete
Digest: sha256:2107a35b58593c58ec5f4e8f2c4a70d195321078aebfadfbfb223a2ff4a4ed21
Status: Downloaded newer image for busybox:latest
PS C:\> docker save -o x:\lcow-save-busybox.tar busybox
PS C:\> dir X:\lcow-save-busybox.tar


    Directory: X:\


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        3/14/2018  12:39 PM        1369600 lcow-save-busybox.tar


PS C:\>
```

And the import on Linux:

image

### CODE REVIEW

**Comment by jhowardmsft** [comment]
@patricklang FYI. Resolves VSO #13274673

**Comment by jhowardmsft** [comment]
@johnstep PTAL.  @gupta-ak @jstarks @darrenstahlmsft FYI.

**Comment by jhowardmsft** [comment]
ping @johnstep

**Comment by thaJeztah** [comment]
Discussing in the maintainers meeting with @tonistiigi and @johnstep and we were wondering why the changes to the storage driver are needed, as `docker push` would also have to get access to these layers/diffs?

Also; this patch would only be temporary until we integrate with the containerd image store.

We're okay with getting this in, just wondering the above :smile:

**Comment by thaJeztah** [comment]
And; would like to see a test :D

**Comment by thaJeztah** [comment]
Guess that depends on https://github.com/moby/moby/pull/34339, correct?

**Comment by jhowardmsft** [comment]
@thaJeztah Push doesn't work yet for LCOW (it might with these changes, but haven't tried). It probably almost certainly goes through the same path. Windows containerd is still some way off yet to be production ready, and it will probably be staged WCOW before LCOW. So needed here in the interim certainly. And yes, can only do manual verification currently.

**Review by thaJeztah — APPROVED** [review]
LGTM, but waiting on @johnstep, and hope we'll have LCOW enabled in CI soon 😅

**Comment by jhowardmsft** [comment]
ping @johnstep

**Comment by kasper3** [comment]
John, two cows are waiting for your review! :)
On a more serious note, if John isn't available, could someone else review this PR.

**Review by johnstep — APPROVED** [review]
LGTM

File: daemon/graphdriver/lcow/lcow.go
Comment (thaJeztah) [reply daemon/graphdriver/lcow/lcow.go]: @johnstep want that to be done as part of this PR, or a follow-up?

Comment (johnstep) [reply daemon/graphdriver/lcow/lcow.go]: It's definitely okay to do it separately from this PR.

**Comment by thaJeztah** [comment]
Janky failure is not related (https://jenkins.dockerproject.org/job/Docker-PRs-experimental/40401/console), and is being tracked through https://github.com/moby/moby/issues/32673

```
21:35:58 FAIL: docker_api_swarm_test.go:296: DockerSwarmSuite.TestAPISwarmLeaderElection
21:35:58 
21:35:58 [df668e46ffa18] waiting for daemon to start
21:35:58 [df668e46ffa18] daemon started
21:35:58 
21:35:58 [d754850ccc130] waiting for daemon to start
21:35:58 [d754850ccc130] daemon started
21:35:58 
21:35:58 [d147e1e5ac54f] waiting for daemon to start
21:35:58 [d147e1e5ac54f] daemon started
21:35:58 
21:35:58 [df668e46ffa18] exiting daemon
21:35:58 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded
21:35:58 [d754850ccc130] exiting daemon
21:35:58 [d147e1e5ac54f] exiting daemon
21:36:22
```

**Pull request merged**
